### PR TITLE
Vertex Color Import/Export, Sharp Normals Fix, and Enhanced Frame Range, Bone Order, and PMX/VMD Export

### PR DESCRIPTION
**Pull request by rintrint**
- Add support for vertex color import/export via ADD UV2
- Fix incorrect custom normals on sharp edges during import
- Add Vertex Splitting option for PMX export
- Better handle cases where users manually adjust frame ranges
- Better handle bone order
- Better handle Bezier interpolation curve
- Add new test script: test_bone_order.py
- Minor code quality improve
- Update README.md

Related Issues:
- #35
- #92
- #120
- #99
- #216
- #219
- #224

# Vertex Colors in MMD Tools

## Overview
MMD Tools supports importing and exporting vertex colors between Blender and PMX format by utilizing the additional UV channels (ADD UV2). The implementation is designed to be compatible with other MMD tools, including minmode's Noesis script vertex color export.

### Exporting from Blender to PMX
- **Vertex colors** in Blender are automatically exported as **ADD UV2** data in PMX format
- The active vertex color layer is converted to RGBA values and stored in the UV2 channel
- Each vertex color component (R, G, B, A) is mapped to the UV2 coordinates (X, Y, Z, W)

### Importing from PMX to Blender
- **ADD UV2** data in PMX files is automatically imported as vertex colors in Blender
- A vertex color layer named "Color" is created to store the color information during import. During export, the active vertex color layer is used regardless of its name
- Additionally, a regular UV layer "UV2" is also created for compatibility

### UV2 Layer Handling
- **When vertex colors are present**: UV layers specifically named "UV2" are treated as converted vertex color data and skipped during export
- **When no vertex colors are present**: UV2 layer is treated as normal additional UV data and exported normally
- This prevents duplication when importing models that already have vertex colors

### With Vertex Colors Present
| Blender Layer | PMX Channel | Description |
|:---------------|:-------------|:-------------|
| UV1 | ADD UV1 | First additional UV layer |
| ~~UV2 (named "UV2")~~ | ~~ADD UV2~~ | Skipped (treated as vertex color data) |
| Vertex Colors (Active Layer) | ADD UV2 | Vertex color data (RGBA) |
| UV3 | ADD UV3 | Third additional UV layer |
| UV4 | ADD UV4 | Fourth additional UV layer |

### Without Vertex Colors
| Blender Layer | PMX Channel | Description |
|:---------------|:-------------|:-------------|
| UV1 | ADD UV1 | First additional UV layer |
| UV2 | ADD UV2 | Second additional UV layer |
| UV3 | ADD UV3 | Third additional UV layer |
| UV4 | ADD UV4 | Fourth additional UV layer |

### Noesis MMD Export Script
MMD Tools vertex color implementation is designed to be compatible with minmode's Noesis script vertex color export:

- **Data Format**: Both use ADD UV2 channel to store RGBA vertex color data
- **Channel Mapping**: R,G,B,A components map to X,Y,Z,W coordinates respectively
- **Import/Export**: Models exported with vertex colors from Noesis script should be compatible with Blender import, and vice versa, provided both tools follow the same ADD UV2 data format
- **Coordinate System**: Vertex color data is stored directly without coordinate transformation, while regular UV data undergoes V-axis flipping for MMD compatibility

This ensures seamless workflow between different MMD modeling tools while maintaining vertex color data integrity.

### Cross-Tool Workflow
1. Export vertex colors from Noesis → Import into Blender (preserves vertex colors)
2. Edit in Blender → Export back to PMX (maintains compatibility)
3. Use in MMD or other tools that support ADD UV2 vertex colors

### UV Layer Limits
- PMX format supports a maximum of **4 additional UV channels**
- If you have more than 3 UV layers + vertex colors, some data will be lost during export
- Priority: Vertex colors are preserved, excess UV layers are discarded

### UV Morph Considerations
- **Manual Management Required**: If you add vertex colors in Blender to a model that originally had UV morphs, you may need to manually adjust the UV morph indices
- UV morphs reference UV channels by index number. When vertex colors occupy the ADD UV2 slot, existing UV layers may be reassigned to different indices, potentially affecting UV morph targeting
- Example: If a UV morph originally targeted UV2 (index 1), after adding vertex colors it should target UV3 (index 2)

### UV2 Layer Behavior
- **Smart Detection**: UV2 layer behavior depends on vertex color presence
- **Automatic Handling**: No manual intervention needed - the script automatically determines whether UV2 should be treated as vertex color data or regular UV data
- **Import Legacy**: Models imported from other tools (like Noesis) create both UV2 layers and vertex color layers from ADD UV2 data, ensuring compatibility with different workflows

### Conflict Resolution
- When both vertex colors and multiple UV layers exist, vertex colors take priority for the ADD UV2 slot
- UV layers are automatically reassigned to available slots (UV1, UV3, UV4)
- Warning messages will inform you of any conflicts or data loss

## Best Practices

1. **Plan ahead**: Decide whether you need vertex colors before creating UV morphs
2. **Check UV morph indices**: After adding vertex colors, verify that UV morphs still target the correct UV layers
3. **Monitor warnings**: Pay attention to console warnings about UV channel conflicts or data loss
4. **Test thoroughly**: Always test your exported model to ensure UV morphs work as expected
5. **Cross-tool compatibility**: When working with models from other tools, import first to check vertex color detection
6. **Layer naming**: Be aware that only UV layers named "UV2" are treated specially when vertex colors are present
7. **Active layer**: Ensure the correct vertex color layer is active before export

## Technical Details

- Vertex colors use 32-bit floating-point precision per channel
- Vertex color data (ADD UV2) does not undergo coordinate flipping, while other UV data requires V-axis flipping to match MMD coordinate system
- ZW channels of additional UVs are stored in separate layers (e.g., "_UV1", "_UV2") when non-zero
- UV2 layer detection is automatic and prevents data duplication during round-trip operations
- Only UV layers specifically named "UV2" are affected by vertex color detection
- The active vertex color layer in Blender is used for export, regardless of layer name
- Vertex color data uses direct RGBA to XYZW mapping without coordinate system conversion

## Console Messages

The script provides informative messages during export:
- `"Vertex colors detected - UV2 layer treated as vertex color data and skipped."`
- `"No vertex colors detected - all UV layers exported normally."`
- These messages help you understand how your UV data is being processed

### CODE REVIEW

**Comment by rintrint** [comment]
I don't think so. MMD Tools already uses `normals_split_custom_set()` to explicitly define per-vertex normals. This strongly indicates that the issue lies within Blender itself, not the data or the add-on.

**Comment by rintrint** [comment]
Blender uses loop normals, which require edges to be marked as sharp to properly separate custom normals. Marking all edges as sharp will create unwanted hard edges, so the correct practice is to mark only the necessary edges as sharp.

I believe Blender assumes that users need to mark sharp edges themselves, so it freely recalculates normals as it sees fit. This behavior seems intentional by design. However, the official documentation does not mention this at all.

**Comment by nagadomi** [comment]
This might be a pretty basic question, but since PMX uses per-vertex normals, is it really necessary to display split normals?
Flat shading can be expressed by actually splitting edges and duplicating vertices.
Even if split/loop normals can be handled in Blender, wouldn't each vertex be limited to only one normal vector when exporting?

**Comment by rintrint** [comment]
This is one of the differences between Blender and other DCCs.
Blender uses loop normals, not vertex normals.
Vertex normals in Blender are always automatically calculated from geometry, so we cannot set them directly.
I originally wanted to use vertex normals too, but Blender does not support that. This is a limitation by design.

Regarding your question: duplicating vertices or splitting geometry just to simulate flat shading is not a good approach.
Preserving the original mesh structure is more important than forcing specific normals, especially when it comes at the cost of introducing unnecessary geometry changes.
Marking edges as sharp is a cleaner and safer way to ensure consistent split normals where needed.
They will have the same values as long as the sharp edges are marked correctly before calling set_custom_split_normals().

File: mmd_tools/core/model.py [hunk 1]
```diff
@@ -47,13 +47,17 @@ def find_armature_object(root_object: bpy.types.Object) -> Optional[bpy.types.Ob
         Returns:
             Optional[bpy.types.Object]: The armature object of the model. If the model does not have an armature, None is returned.
         """
+        if root_object is None:
```
Comment (UuuNyaa) [inline mmd_tools/core/model.py:50]: If the argument could be `None`, it is better to declare the type as `Optional[bpy.types.Object]`.

Comment (rintrint) [reply mmd_tools/core/model.py:50]: Fixed

**Comment by nagadomi** [comment]
>Add Vertex Splitting option for PMX export

I posted my opinion at https://github.com/MMD-Blender/blender_mmd_tools/issues/224#issuecomment-2925231555

File: mmd_tools/core/model.py [hunk 2]
```diff
@@ -40,20 +40,24 @@ def find_root_object(obj: Optional[bpy.types.Object]) -> Optional[bpy.types.Obje
...
+            return None
         for o in root_object.children:
             if o.type == "ARMATURE":
                 return o
         return None
 
     @staticmethod
-    def find_rigid_group_object(root_object: bpy.types.Object) -> Optional[bpy.types.Object]:
+    def find_rigid_group_object(root_object: Optional[bpy.types.Object]) -> Optional[bpy.types.Object]:
+        if root_object is None:
```
Comment (UuuNyaa) [inline mmd_tools/core/model.py:59]: If the argument could be `None`, it is better to declare the type as `Optional[bpy.types.Object]`.

File: mmd_tools/core/model.py [hunk 3]
```diff
@@ -71,41 +75,53 @@ def __new_group_object(context: bpy.types.Context, name: str, mmd_type: str, par
 
     @staticmethod
     def ensure_rigid_group_object(context: bpy.types.Context, root_object: bpy.types.Object) -> bpy.types.Object:
+        if root_object is None:
```
Comment (UuuNyaa) [inline mmd_tools/core/model.py:78]: If the argument could be `None`, it is better to declare the type as `Optional[bpy.types.Object]`.

File: mmd_tools/core/model.py [hunk 4]
```diff
@@ -71,41 +75,53 @@ def __new_group_object(context: bpy.types.Context, name: str, mmd_type: str, par
...
+            return None
         for o in root_object.children:
             if o.type == "EMPTY" and o.mmd_type == "JOINT_GRP_OBJ":
                 return o
         return None
 
     @staticmethod
     def ensure_joint_group_object(context: bpy.types.Context, root_object: bpy.types.Object) -> bpy.types.Object:
+        if root_object is None:
```
Comment (UuuNyaa) [inline mmd_tools/core/model.py:96]: If the argument could be `None`, it is better to declare the type as `Optional[bpy.types.Object]`.

File: mmd_tools/core/model.py [hunk 5]
```diff
@@ -71,41 +75,53 @@ def __new_group_object(context: bpy.types.Context, name: str, mmd_type: str, par
...
+            return None
         for o in root_object.children:
             if o.type == "EMPTY" and o.mmd_type == "TEMPORARY_GRP_OBJ":
                 return o
         return None
 
     @staticmethod
     def ensure_temporary_group_object(context: bpy.types.Context, root_object: bpy.types.Object) -> bpy.types.Object:
+        if root_object is None:
```
Comment (UuuNyaa) [inline mmd_tools/core/model.py:114]: If the argument could be `None`, it is better to declare the type as `Optional[bpy.types.Object]`.

File: mmd_tools/core/vmd/exporter.py [hunk 6]
```diff
@@ -72,12 +72,12 @@ def __toVMDControlPoints(bezier):
         dx, dy = p3 - p0
         x1, y1 = p1 - p0
         x2, y2 = p2 - p0
-        if abs(dx) < 5e-5:
+        if abs(dx) == 0.0:
```
Comment (UuuNyaa) [inline mmd_tools/core/vmd/exporter.py]: Have you considered using `math.isclose`?

Comment (rintrint) [reply mmd_tools/core/vmd/exporter.py]: Thanks for the suggestion. I tried many very small thresholds, even as small as < 1e-323, but after testing, I found that checking == 0.0 gave the most consistent and accurate behavior.

Based on my testing, the calculation should always run unless it would result in division by zero. So using an exact zero check seems to be the best fit here.

File: mmd_tools/core/pmx/exporter.py [hunk 7]
```diff
@@ -882,38 +897,102 @@ def __exportJoints(self, joints, rigid_map):
             p_joint.spring_rotation_constant = Vector(mmd_joint.spring_angular).xzy
             self.__model.joints.append(p_joint)
 
-    @staticmethod
-    def __convertFaceUVToVertexUV(vert_index, uv, normal, vertices_map):
+    def __convertFaceUVToVertexUV(self, vert_index, uv, normal, vertices_map, face_area):
```
[the remark this answers is not in the document]
Comment (rintrint) [reply mmd_tools/core/pmx/exporter.py]: Fixed

File: mmd_tools/core/vmd/exporter.py (hunk 6 above)
Comment (UuuNyaa) [reply mmd_tools/core/vmd/exporter.py]: Let us believe the test results!

Comment (rintrint) [reply mmd_tools/core/vmd/exporter.py]: I'm still investigating this issue in more depth.
Setting the threshold to == 0.0 does reduce the number of bones that fail in test_vmd_exporter.py, but unfortunately, it doesn't eliminate them completely.

Comment (rintrint) [reply mmd_tools/core/vmd/exporter.py]: I've relaxed the conditions further after additional investigation. When dy is very small, the handle positions become essentially meaningless for animation purposes.

The current approach now uses a more practical threshold that accounts for the reality that tiny value changes don't produce visually significant animation differences, and trying to preserve exact handle positions in these cases often introduces more artifacts than benefits.

This change prioritizes animation quality and import/export consistency over mathematical precision in edge cases where the precision doesn't contribute to meaningful motion.

**Comment by rintrint** [comment]
All test cases are now passing. Requirements for PMX files validity:
1. No unreferenced vertices
2. No NaN values in display connections
3. Exported by MMD Tools (morph order issue not yet fixed)

Snipaste_2025-06-05_05-23-17

**Comment by rintrint** [comment]
@UuuNyaa Please take a look when you have time. I believe this is ready to merge and we can proceed with the release.

**Comment by rintrint** [comment]
All default Ruff rules now pass using the following command:
```
ruff check
```
Or, to check against specific selected rules:
```
ruff check --select I,F401,UP009,Q,W291,D412,W293,F541,E402,E713,E701,F507,E722,F821,E731,E741,F841 --line-length 320
I       isort
F401    unused import
UP009   UTF-8 encoding declaration
Q       flake8-quotes
W291    trailing whitespace
D412    blank lines between header and content
W293    blank line with whitespace
F541    f-string missing placeholders (remove extraneous `f` prefix)
E402    module-level import not at top of file
E713    use `not in` instead of `not {x} in {y}`
E701    multiple statements on one line
F507    percent-format positional count mismatch
E722    bare `except`
F821    undefined name
E731    do not assign a lambda expression, use `def` instead
E741    ambiguous variable name
F841    local variable is assigned but never used
--line-length 320 - Uses a very permissive line length to minimize code changes
```

File: mmd_tools/core/vmd/importer.py [hunk 8]
```diff
@@ -358,8 +409,13 @@ def compatible_rotation(prev, curr):
                     return (angle, x, y, z)
 
             else:
-                convert_rotation = lambda rot: converter.convert_rotation(rot).to_euler(mode)
-                compatible_rotation = lambda prev, curr: curr.make_compatible(prev) or curr
+                @staticmethod
+                def convert_rotation(rot):
```
[the remark this answers is not in the document]
Comment (rintrint) [reply mmd_tools/core/vmd/importer.py:413]: Thanks for the feedback. Won't fix due to:
1. If it ain't broke, don't fix it
2. I will fix code issues if and only if the messy code becomes a blocker for solving new issues
3. Out of this PR scope. I only solve "Fix Ruff E731 do not assign a lambda expression, use a def" - this problem existed before
